File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        stack<int> st;
        int ans = 0;

        for(int i = 0; i<n; i++) {
            while(!st.empty() && height[st.top()] < height[i]) {
                int t = st.top();
                st.pop();
                if(st.empty()) break;
                int h = min(height[st.top()], height[i]) - height[t];
                int l = i - st.top() - 1;
                ans += (l*h);
            }
            st.push(i);
        }

        return ans;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        n = len(height)
        stack = []
        ans = 0

        for i in range(n):
            while stack and height[stack[-1]] < height[i]:
                t = stack[-1]
                stack.pop()
                if len(stack) == 0:
                    break
                h = min(height[i], height[stack[-1]]) - height[t]
                l = i - stack[-1] - 1
                ans += (l * h)
            stack.append(i)

        return ans

        